File: PixelSolution/PixelEngine/Debug.cpp
#include "pch.h"
#include "Debug.h"
#include "PixelGraphicsAPI.h"

void Debug_Log(const char* log)
{
	std::string logSTR(log);
	PixelLog::Info(logSTR);
}

void Debug_LogError(const char* log)
{
	std::string logSTR(log);
	PixelLog::Error(logSTR);
}

void Debug_LogWarning(const char* log)
{
	std::string logSTR(log);
	PixelLog::Warn(logSTR);
}

void Debug_Line(PVector3 start, PVector3 end, PVector3 color)
{
	constexpr float INV_255 = 1.0f / 255.0f;
	DrawLine(start.x, start.y, start.z, end.x, end.y, end.z, color.x * INV_255, color.y * INV_255, color.z * INV_255);
}

void Debug_Box2D(float x, float y, float z, float width, float height, float r, float g, float b)
{
	constexpr float INV_255 = 1.0f / 255.0f;
	float colorR = r * INV_255;
	float colorG = g * INV_255;
	float colorB = b * INV_255;
	float halfX = width * 0.5f;
	float halfY = height * 0.5f;
	float left = x - halfX;
	float right = x + halfX;
	float bottom = y - halfY;
	float top = y + halfY;
	DrawLine(left, bottom, z, right, bottom, z, colorR, colorG, colorB);
	DrawLine(right, bottom, z, right, top, z, colorR, colorG, colorB);
	DrawLine(right, top, z, left, top, z, colorR, colorG, colorB);
	DrawLine(left, top, z, left, bottom, z, colorR, colorG, colorB);
}

void Debug_Sphere2D(float x, float y, float z, float radius, float r, float g, float b)
{
	constexpr float INV_255 = 1.0f / 255.0f;
	DrawCircle2D(x, y, z, radius, r * INV_255, g * INV_255, b * INV_255);
}

File: PixelSolution/PixelEngine/Debug.lua
local ffi = require("ffi")

ffi.cdef[[
    void Debug_Log(const char* log);
    void Debug_LogError(const char* log);
    void Debug_LogWarning(const char* log);

    void Debug_Line(float startX, float startY, float startZ, float endX, float endY, float endZ, float r, float g, float b);
    void Debug_Box2D(float x, float y, float z, float width, float height, float r, float g, float b);
    void Debug_Sphere2D(float x, float y, float z, float radius, float r, float g, float b);
]]

local dll = ffi.load("PixelEngine")

Debug = Debug or {}

---@param log string
function Debug.Log(log)
    dll.Debug_Log(tostring(log))
end

---@param log string
function Debug.LogError(log)
    dll.Debug_LogError(tostring(log))
end

---@param log string
function Debug.LogWarning(log)
    dll.Debug_LogWarning(tostring(log))
end

---@param startX number
---@param startY number
---@param startZ number
---@param endX number
---@param endY number
---@param endZ number
---@param r number
---@param g number
---@param b number
function Debug.Line(startX, startY, startZ, endX, endY, endZ, r, g, b)
    dll.Debug_Line(startX, startY, startZ, endX, endY, endZ, r, g, b)
end

---@param x number
---@param y number
---@param z number
---@param width number
---@param height number
---@param r number
---@param g number
---@param b number
function Debug.Box2D(x, y, z, width, height, r, g, b)
    dll.Debug_Box2D(x, y, z, width, height, r, g, b)
end

---@param x number
---@param y number
---@param z number
---@param radius number
---@param r number
---@param g number
---@param b number
function Debug.Sphere2D(x, y, z, radius, r, g, b)
    dll.Debug_Sphere2D(x, y, z, radius, r, g, b)
end

return Debug
